extjson.go

```diff
@@ -0,0 +1,1648 @@
+// $dbPointer
+// $numberInt
+// $timestamp
+// $undefined
+// $numberLong
+// $numberDouble
+// $numberDecimal
+// $regularExpression
+
+// handleExtJSON is called from convertObject to potentially replace a JSON
+// object with a non-document BSON value instead.  If it returns (nil, nil), it
+// means that the input is not extended JSON and that no bytes were consumed
+// from the input.  If it succeeds, then the extended JSON object will have been
+// consumed.
+//
+// If/when a type is definitively determined, the `typeBytePos` of
+// `out` will be overwritten the discovered type.
+//
+// This function generally only validates/dispatches to subroutines.  Those
+// functions are responsible for consuming the full extended JSON object,
+// including the closing terminator.
+func (d *Decoder) handleExtJSON(out []byte, typeBytePos int) ([]byte, error) {
+	// Peek ahead for longest possible extjson key plus closing quote.
+	buf, err := d.json.Peek(19)
+	if err != nil {
+		// May have peeked to end of input, so EOF is OK.
+		if err != io.EOF {
+			return nil, err
+		}
+	}
+
+	// Common case: not extended JSON.
+	if len(buf) > 0 && buf[0] != '$' {
+		return nil, nil
+	}
+
+	// Isolate key
+	quotePos := bytes.IndexByte(buf, '"')
+	if quotePos < 0 {
+		// Key is longer than `$regularExpression"`, so not valid extended JSON.
+		return nil, nil
+	}
+	key := buf[0:quotePos]
+
+	// When we find a key, we can write a type byte and discard from the input
+	// buffer the length of that key plus one for the closing quote.  In
+	// ambiguous cases, we can't assign a type or discard, so we defer that
+	// to a corresponding subroutine.
+	switch len(key) {
+	case 4: // $oid
+		if bytes.Equal(key, jsonOID) {
+			overwriteTypeByte(out, typeBytePos, bsonObjectID)
+			_, _ = d.json.Discard(5)
+			return d.convertOID(out)
+		}
+		return nil, nil
+	case 5: // $code $date $type
+		if bytes.Equal(key, jsonCode) {
+			// Still don't know if this is code or code w/scope, so can't
+			// assign type yet, but we can consume the key.
+			_, _ = d.json.Discard(6)
+			return d.convertCode(out, typeBytePos)
+		} else if bytes.Equal(key, jsonDate) {
+			overwriteTypeByte(out, typeBytePos, bsonDateTime)
+			_, _ = d.json.Discard(6)
+			return d.convertDate(out)
+		} else if bytes.Equal(key, jsonType) {
+			// Still don't know if this is binary or a $type query operator, so
+			// can't assign type or discard anything yet.
+			return d.convertType(out, typeBytePos)
+		}
+		return nil, nil
+	case 6: // $scope $regex
+		if bytes.Equal(key, jsonScope) {
+			overwriteTypeByte(out, typeBytePos, bsonCodeWithScope)
+			_, _ = d.json.Discard(7)
+			return d.convertScope(out)
+		} else if bytes.Equal(key, jsonRegex) {
+			// Still don't know if this is legacy $regex or a $regex query
+			// operator so can't assign type or discard anything yet.
+			return d.convertRegex(out, typeBytePos)
+		}
+		return nil, nil
+	case 7: // $binary $maxKey $minKey $symbol
```

Comment:
So we'd definitely consider this case for byte by byte switching for early exit.

Reply:
Also, since these are 7 bytes AND there are more else's, I would expect this particular case to yield more fruit.